File: src/main/java/recipestore/ResourceLoader.java
package recipestore;

import lombok.RequiredArgsConstructor;
import org.jooq.lambda.Seq;
import org.jooq.lambda.tuple.Tuple2;

import java.util.Map;
import java.util.Optional;
import java.util.ResourceBundle;
import java.util.function.BiFunction;
import java.util.stream.Collectors;

import static java.util.Optional.ofNullable;

public class ResourceLoader {
    final static Map<Resource, ResourceBundle> resourceCache;

    static {
        resourceCache = Seq.of(Resource.values())
                .map(resource -> new Tuple2<>(resource, ResourceBundle.getBundle(resource.resourceName)))
                .collect(Collectors.toMap(tuple -> tuple.v1, tuple -> tuple.v2));
    }

    @RequiredArgsConstructor
    public enum Resource {
        triplestore("triplestore"), graph("graph"), lucene("lucene");
        private final String resourceName;

        @Override
        public String toString() {
            return resourceName;
        }
    }

    public static BiFunction<Resource, String, Optional<String>> get = (resource, key) ->
            ofNullable(resourceCache.get(resource).containsKey(key) ?
                    resourceCache.get(resource).getString(key)
                    : null);

}

File: src/main/scala/recipestore/ResourceLoader.scala
package recipestore

import java.util.ResourceBundle

object AppResource extends Enumeration {
  type ResourceName = Value
  val TinkerpopResource = Value("tinkerpop")
  val TriplestoreResource = Value("triplestore")
  val GraphResource = Value("graph")
  val LuceneResource = Value("lucene")
}

object ResourceLoader {
  private def resourceBundle(key: String) = ResourceBundle.getBundle(key)

  val resourceCache: Map[AppResource.Value, ResourceBundle] =
    Map(
      AppResource.TinkerpopResource -> resourceBundle(AppResource.TinkerpopResource.toString),
      AppResource.TriplestoreResource -> resourceBundle(AppResource.TriplestoreResource.toString),
      AppResource.GraphResource -> resourceBundle(AppResource.GraphResource.toString),
      AppResource.LuceneResource -> resourceBundle(AppResource.LuceneResource.toString)
    )


  def apply(resource: AppResource.Value, key: String): Option[String] = {
    val resourceBundle = resourceCache.getOrElse(resource, null)
    if (resourceBundle != null && resourceBundle.containsKey(key)) Option(resourceBundle.getString(key))
    else Option.empty
  }


}